### PR TITLE
411: prevent flush entry followed delete operations

### PR DESCRIPTION
**Pull request by ayulas**

### CODE REVIEW

File: db/memtable.cc [hunk 1]
```diff
@@ -737,7 +737,9 @@ Status MemTable::Add(SequenceNumber s, ValueType type,
                                val_size + moptions_.protection_bytes_per_key;
   char* buf = nullptr;
   std::unique_ptr<MemTableRep>& table =
-      type == kTypeRangeDeletion ? range_del_table_ : table_;
+      type == kTypeRangeDeletion /*||
```
Comment (mrambacher) [inline db/memtable.cc:740]: Is this leftover debugging code?

File: include/rocksdb/options.h [hunk 2]
```diff
@@ -1181,6 +1181,10 @@ struct DBOptions {
   // Not supported in ROCKSDB_LITE mode!
   std::shared_ptr<Cache> row_cache = nullptr;
 
+  // If true during flush we skip any entry that has a followed single delete entry also in 
+  // the not flushed mutable and immutable memtables  (#411)
+  bool use_clean_single_delete_during_flush = false;
+
```
Comment (mrambacher) [inline include/rocksdb/options.h:1187]: I am not sure I understand what the comment is trying to say.  

The variable name seems very long and I am not sure what the words together mean.  Why "clean"?  Why not "skip_single_delete_during_flush", or something like that?

File: options/db_options.cc [hunk 3]
```diff
@@ -570,6 +570,11 @@ static std::unordered_map<std::string, OptionTypeInfo>
          {offsetof(struct ImmutableDBOptions, use_dynamic_delay),
           OptionType::kBoolean, OptionVerificationType::kNormal,
           OptionTypeFlags::kNone}},
+        {"use_clean_single_delete_during_flush",
+         {offsetof(struct ImmutableDBOptions,
+                   use_clean_single_delete_during_flush),
+          OptionType::kBoolean, OptionVerificationType::kNormal,
+          OptionTypeFlags::kNone}},
```
Comment (mrambacher) [inline options/db_options.cc:577]: Is there a reason this option should be immutable?

File: include/rocksdb/options.h [hunk 4]
```diff
@@ -1181,6 +1181,10 @@ struct DBOptions {
   // Not supported in ROCKSDB_LITE mode!
   std::shared_ptr<Cache> row_cache = nullptr;
 
+  // If true during flush we skip any entry that has a followed single delete entry also in 
+  // the not flushed mutable and immutable memtables  (#411)
+  bool use_clean_single_delete_during_flush = false;
```
Comment (mrambacher) [inline include/rocksdb/options.h:1186]: I do not see where this option is used in the code.  Is this specific to a certain memtable or am I missing its use?

File: options/db_options.cc (hunk 3 above)
Comment (ayulas) [reply options/db_options.cc:577]: yes it should be as other delete options

File: include/rocksdb/options.h (hunk 2 above)
Comment (ayulas) [reply include/rocksdb/options.h:1187]: becuase its clean entry that has single delete

File: include/rocksdb/options.h (hunk 4 above)
Comment (ayulas) [reply include/rocksdb/options.h:1186]: this was set in the first phase and. it was decided we want it any way. so i should remove it probably. but i kept it cause i wasnt sure i wont want to use it

**Review by hilikspdb — DISMISSED** [review]

File: db/builder.cc [hunk 5]
```diff
@@ -205,16 +205,53 @@ Status BuildTable(
...
       const Slice& value = c_iter.value();
       const ParsedInternalKey& ikey = c_iter.ikey();
       // Generate a rolling 64-bit hash of the key and values
       // Note :
       // Here "key" integrates 'sequence_number'+'kType'+'user key'.
+      if (ioptions.use_clean_delete_during_flush &&
+          tboptions.level_at_creation == 0 && ikey.type == kTypeValue) {
+        bool was_skipped = false;
+        while (range_del_it->Valid()) {
+          auto internal_key = InternalKey(key, ikey.sequence, ikey.type);
```
Comment (mrambacher) [inline db/builder.cc:232]: What is the difference between ikey and internal_key?

Comment (mrambacher) [inline db/builder.cc:232]: If needed, this internal_key should be done outside of the while loop (I do not see the parameters changing inside this loop)

File: db/compaction/compaction_iterator.cc [hunk 6]
```diff
@@ -358,6 +359,71 @@ bool CompactionIterator::InvokeFilterIfNeeded(bool* need_skip,
...
 }
 
+bool CompactionIterator::CanBeSkipped() {
+  if (!use_skip_delete_) {
+    return false;
+  }
+  key_ = input_.key();
+  value_ = input_.value();
+
+  Status pik_status = ParseInternalKey(key_, &ikey_, allow_data_in_errors_);
```
Comment (mrambacher) [inline db/compaction/compaction_iterator.cc:369]: Where is pik_status checked?

**Review by mrambacher — COMMENTED** [review]
Do you have unit tests to cover these changes?

File: db/memtable.cc [hunk 7]
```diff
@@ -736,8 +736,11 @@ Status MemTable::Add(SequenceNumber s, ValueType type,
                                internal_key_size + VarintLength(val_size) +
                                val_size + moptions_.protection_bytes_per_key;
   char* buf = nullptr;
-  std::unique_ptr<MemTableRep>& table =
-      type == kTypeRangeDeletion ? range_del_table_ : table_;
+  std::unique_ptr<MemTableRep>& table = type == kTypeRangeDeletion /*||
+                                        type == kTypeSingleDeletion ||
+                                        type == kTypeDeletion*/
```
[the remark this answers is not in the document]
Comment (ayulas) [reply db/memtable.cc]: this comment is an old commit it doesnt existin the code

File: db/builder.cc (hunk 5 above)
Comment (ayulas) [reply db/builder.cc:232]: // The data structure that represents an internal key in the way that user_key,
// sequence number and type are stored in separated forms.
struct ParsedInternalKey

**Review by hilikspdb — DISMISSED** [review]

**Review by hilikspdb — DISMISSED** [review]

**Review by hilikspdb — APPROVED** [review]
i think this option should be CF option but for now I approve ....

**Review by Yuval-Ariel — DISMISSED** [review]

File: options/options_helper.cc [hunk 8]
```diff
@@ -186,10 +186,13 @@ DBOptions BuildDBOptions(const ImmutableDBOptions& immutable_db_options,
   options.checksum_handoff_file_types =
       immutable_db_options.checksum_handoff_file_types;
   options.lowest_used_cache_tier = immutable_db_options.lowest_used_cache_tier;
+  options.use_dynamic_delay = immutable_db_options.use_dynamic_delay;
```
Comment (Yuval-Ariel) [inline options/options_helper.cc:189]: no reason to add this. its a dup of line 139

**Comment by ayulas** [comment]
@Yuval-Ariel can you pls review the db_stress db_crash changes and approve so it will go to testing? tnx

File: options/options_helper.cc (hunk 8 above)
Comment (ayulas) [reply options/options_helper.cc]: done

**Comment by Yuval-Ariel** [comment]
LGTM
btw, its very hard to see only the recent changes since you force pushed after rebasing. its easier to see the changes without force pushing and at the end we can squash and merge.

**Review by Yuval-Ariel — APPROVED** [review]

**Review by Yuval-Ariel — DISMISSED** [review]

**Comment by ayulas** [comment]
Black box test failed  I’m in it

**Review by hilikspdb — APPROVED** [review]

**Comment by Yuval-Ariel** [comment]
need to update the history file

**Comment by ayulas** [comment]
history file updated and also @Yuval-Ariel remark to me

**Comment by ayulas** [comment]
ci https://github.com/speedb-io/speedb/actions/runs/5236684706 passed
@Yuval-Ariel pls merge

**Pull request merged**